### DISCUSSION TITLE
conditional validation with field set

### DISCUSSION THREAD

**Discussion by niralparmar** [react-hook-form/react-hook-form#3657] [Q&A] [answered]
hi,
in my form i have one field set and within couple input fields (just for the simplicity , but it could have complex nested form component).
requirement here is when a checkbox get selected , input within should be marked as required when uncheck it should not be marked required and its data should be removed from formData if any data enters by user.

here is my [codesandbox](https://codesandbox.io/s/react-hook-form-conditional-fields-forked-mj7sl?file=/src/index.js)

how can i achieve this?

thank you

**Comment by bluebill1049** [comment]
use `validate` with `getValues`

**Reply by niralparmar** [reply]
thanks @bluebill1049 , validate gets fired when you click submit , 
 in my use case, submit button is disabled untill form is valid. there is one field outside the fiedset which is required, if user enters a value form state should be valid because if fieldset in unchecked. 
 i have updated my codesandbox for more clarity 

appreciate your help

**Reply by bluebill1049** [reply]
is this dependant validation, you can use `trigger` to trigger the other input's validation.

**Reply by niralparmar** [reply]
yes it depends on "checkbox" , also if "checkbox" is unchecked those fields should be not part of formData.
initially "first" and "last" will not be marked as required, even if i trigger, those fields will not get validated.

main issue i see here, how can i dynamically register/unregister fields based on certain fields.  how can i incorporate register/unregister with regards to "fieldset"

**Reply by niralparmar** [reply]
@bluebill1049  even i added this code (update my codesanbox as well) but form doesnt work as expected
`useEffect(() => {
    // @ts-ignore

    if (!disabled) {
      console.log("unregister");
      unregister("firstName");
      unregister("lastname");
    } else {
      console.log("register");
      register("firstName", { required: true });
      register("lastName", { required: true });
    }
  }, [disabled, register, unregister]);`

**Comment by bluebill1049** [comment]
if you can "naturally" let a component unmount, then register will remove that input from submission.

`{xxx && <input ref={register} />} // when xxx become false, the input will get removed from submission`

**Reply by niralparmar** [reply]
thanks @bluebill1049 for the reply.. that would mean component will get hidden from ui , which is not the way i wanted.
if you look at my codesand box, after register form doesnt recognize that i have already  entered some values. 

![Uploading Screen Shot 2020-12-12 at 11.53.09 PM.png…]()

**Reply by niralparmar** [reply]
@bluebill1049 i am in real need of help here....as all of our forms has this kind of requirement ,having conditional and dynamic validation.

thank you

**Reply by niralparmar** [reply]
i tried this simple register/unregister from your example , even that is not working. i am not sure what i am missing
[codesandbox](https://codesandbox.io/s/react-hook-form-unregister-v6-forked-19446?file=/src/index.js:553-562)

**Reply by bluebill1049** [reply]
if you not even `setValue`, the validation obviously block the onSubmit: https://codesandbox.io/s/react-hook-form-unregister-v6-forked-wm9j1?file=/src/index.js

**Reply by niralparmar** [reply]
thanks @bluebill1049  so i manually register a field, i have to explicitly call 'setValue`?   because i still dont see data getting submitted in your [csb]( https://codesandbox.io/s/react-hook-form-unregister-v6-forked-wm9j1?file=/src/index.js)
this code simply wont work than,

> useEffect(() => {
    register("firstName", { required: true });
    register("lastName", { required: true });
    // }
  }, [register]);

**Reply by niralparmar** [reply]
@bluebill1049 i have updated my original issue [csb](https://codesandbox.io/s/react-hook-form-conditional-fields-forked-mj7sl?file=/src/index.js) with `setValue` upon `onChange`, but formdata doesnt get updated.  and form stays in invalid state

**Reply by bluebill1049** [reply]
i can see you are frustrated... I will send it sometime during my lunch break to have a look at your question. saw that on stackoverflow as well.

https://stackoverflow.com/questions/65257788/conditional-validation-with-react-hook-form

(i didn't down vote)

**Reply by bluebill1049** [reply]
This whole validation behavior (UX) is definitely making things a bit harder, however, there are a couple of things that you should leverage from the library such as: 

 - `watch`
 - `validate`
 - `getValues`

``` javascript
import React from "react";
import ReactDOM from "react-dom";
import { useForm } from "react-hook-form";

import "./index.css";

function App() {
  const {
    register,
    handleSubmit,
    errors,
    formState: { isValid, touched },
    getValues,
    trigger,
    watch
  } = useForm({
    mode: "onBlur"
  });
  const onSubmit = (data) => {
    alert(JSON.stringify(data));
  };
  const validate = (value) => {
    if (getValues("name")) { // read the checkbox value
      return !!value;
    }

    return true;
  };
  const isChecked = watch("name"); // watch if the name is checked

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <label htmlFor="id">ID</label>
      <input
        name="id"
        placeholder="id"
        ref={register({ required: true, maxLength: 50 })}
      />
      {errors.id && <p>"ID is required"</p>}

      <fieldset disabled={!isChecked}>
        <legend>
          <input
            type="checkbox"
            name={"name"}
            ref={register}
            onChange={() => trigger()} // you want update isValid due to state change, and also those extra two inputs become required
          />
          <span>Check</span>
        </legend>
        <label htmlFor="firstName">First Name</label>
        <input
          name="firstName"
          placeholder="Bill"
          ref={register({
            validate
          })}
        />
        // make sure input is touched before fire an error message to the user
        {errors.firstName && touched["firstName"] && (
          <p>"First name is required"</p>
        )}
        <label htmlFor="lastName">Last Name</label>
        <input
          name="lastName"
          placeholder="Luo"
          ref={register({
            validate
          })}
        />
        {errors.lastName && touched["lastName"] && (
          <p>"Last name is required"</p>
        )}
      </fieldset>

      <input type="submit" disabled={!isValid} />
    </form>
  );
}

const rootElement = document.getElementById("root");
ReactDOM.render(<App />, rootElement);
```

CSB: 
https://codesandbox.io/s/react-hook-form-conditional-fields-forked-n0jig?file=/src/index.js:0-1831

**Reply by niralparmar** [reply]
@bluebill1049  much appreciate your help. i really need help in here. this is one of the most common and important use case we are trying and i really want this library to work.
i looked at your code, one use case is still failing (i am pushing my luck here) , if user enters data in "first" and "last" name than unchecked the checkbox , data still remains in the formData, it should also get removed.

but also, i am trying to understand why register/unregister is not working. seems those api's are meant for such use cases.

again, many thank you for your prompt reply.

**Reply by bluebill1049** [reply]
I think it makes sense to remove at submitting, otherwise you are making this too hard for yourself (unless this is a business requirement)

**Reply by niralparmar** [reply]
it is a business requirement. but thank you so much putting extra effort into this issue. (but i am surprise this is not of of the common scenario)

**Reply by bluebill1049** [reply]
This is not common, because, for controlled form, the state of the input remains the same over the course of the form. React Hook Form is actually doing extra work to remove input state when input is **unmounted**, which you don't even have that with controlled input, but with your case, it's tricky because your input remains registered while you want to remove then, that's why it's I recommend you doing that transform at submitting callback.

**Reply by niralparmar** [reply]
thats is why i was looking forward if "unregister" could do this trick..

**Reply by bluebill1049** [reply]
`unregister` should work, you can follow what I have above (but not register those inputs at `ref`) obviously more extra work, you need to use setValue with shouldValdiate and etc...

**Comment by niralparmar** [comment, marked as answer]
[csb](https://codesandbox.io/s/react-hook-form-conditional-fields-forked-n0jig?file=/src/index.js:0-1831) provided by @bluebill1049 works after some modification.
